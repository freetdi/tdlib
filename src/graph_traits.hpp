// Felix Salfelder, 2016
//
// (c) 2016 Goethe-Universit√§t Frankfurt
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//
//   traits for tdlib graphs and treedecs.
//
//

#ifndef TD_GRAPH_TRAITS_HPP
#define TD_GRAPH_TRAITS_HPP

#include <boost/graph/graph_traits.hpp>
#include <boost/graph/adjacency_list.hpp>
#include <set>

namespace treedec{
#ifndef TD_DEFS_NETWORK_FLOW
#define TD_DEFS_NETWORK_FLOW

struct Vertex_NF{
    bool visited;
    int predecessor;
};

struct Edge_NF{
    bool path; //true if a path uses the edge
};

#endif

// ouch. this is actually bag_t
// how to fix that now?
// "bag" is also used as shorthand for bag access...

struct bag_t{ //
    std::set<unsigned int> bag;
};

// dont define twice (in old code)
#define TD_STRUCT_BAG

}// treedec

// KLUGE: put it here...
// (and cross fingers)
//using bag = treedec::bag_t;

namespace treedec{


template<class G_t>
struct graph_traits_base : public boost::graph_traits<G_t> {};
// kludge for balu
// TODO: use graph_traits. see below
// namespace detail{ not yet
template<class G>
struct treedec_chooser{ //
    typedef unsigned value_type;
    typedef std::set<unsigned> bag_type;
    typedef boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, bag_t> type;
};
// } not yet

// this makes some sense...
template<class G_t>
struct graph_traits : public graph_traits_base<G_t> { //
    typedef typename treedec_chooser<G_t>::type treedec_type;
    typedef typename std::set<unsigned> outedge_set_type;
    typedef typename boost::adjacency_list<boost::vecS, boost::vecS,
                boost::bidirectionalS, Vertex_NF, Edge_NF> directed_overlay;
    typedef typename boost::adjacency_list<boost::setS, boost::vecS, boost::undirectedS> immutable_type;
};

// obsolete. use graph_traits directly.
template<class G>
struct outedge_set{
    typedef typename graph_traits<G>::outedge_set_type type;
};

// test if v is a valid vertex_descriptor of g
template<typename G>
inline bool is_valid(typename boost::graph_traits<G>::vertex_iterator const &,
		const G&)
{
    return true;
}
template<typename G>
inline bool is_valid(typename boost::graph_traits<G>::vertex_descriptor const &,
		const G&)
{
    return true;
}


namespace detail{ //
// working around balu and bag
template <bool, typename T = void>
struct vdstuff { //
    typedef unsigned type;
    typedef std::set<unsigned> bag_type;
};

// temporary hack don't touch.
// specialize teedec_traits (below) in case you must.
template <typename T>
struct vdstuff<false, T> { //
    typedef typename T::value_type type;
    typedef T bag_type;
};
} //detail

template<class T>
struct treedec_traits{ //
// TODO should be this (does not work, why?)
//    typedef typename boost::graph_traits<T>::vertex_property_type vertex_property_type;
    typedef typename T::vertex_property_type vertex_property_type;
    typedef typename detail::vdstuff<
       boost::is_same<vertex_property_type, bag_t >::value,
         vertex_property_type >::type vd_type;

    typedef typename detail::vdstuff<
       boost::is_same<vertex_property_type, bag_t >::value,
         vertex_property_type >::bag_type bag_type;

};

}

// return "id" where the vertex_descriptor might make more sense.
// (transitional interface)
template<typename G>
inline unsigned get_vd(const G&, const typename boost::graph_traits<G>::vertex_descriptor& v )
{
    // works with "TD_graph_t" (augmented adj_list)
    //return g[v].id;
    return v;
}

//Return the internal vertex position.
//To be used as a narrower alternative to vertex_descriptor.
//Positions are in {0, 1, ..., num_vertices-1}, where applicable.
//(One you use the vertex descriptor in boost graphs with vertex container 'vecS').
// this position must be stable under copy and assignment operations.

namespace treedec{

template<typename G_t>
inline typename boost::graph_traits<G_t>::vertices_size_type
   get_pos(typename boost::graph_traits<G_t>::vertex_descriptor v, G_t const& G)
{
    return boost::get(boost::vertex_index, G, v);
}

} // treedec

namespace treedec{

// chooose deg implementation for graph backend.
// to be accessed through graph_traits
template<class G_t>
struct deg_chooser;

template<class G>
void check(G const&)
{
}

namespace detail{
  template<class G>
  class shared_adj_iter;
}

template<typename vertex_descriptor>
struct vertex_callback{ //
    virtual ~vertex_callback(){};
    virtual void operator()(vertex_descriptor)=0;
};

template<typename G_t>
struct edge_callback{ //
    typedef typename boost::graph_traits<G_t>::edge_descriptor edge_descriptor;
    typedef typename boost::graph_traits<G_t>::vertex_descriptor vertex_descriptor;
    virtual ~edge_callback(){};
    virtual void operator()(vertex_descriptor, vertex_descriptor)=0;
    void operator()(edge_descriptor)
    { incomplete();
    }
};

template<typename G_t>
struct graph_callback{ // fixme: union of the above?
    typedef typename boost::graph_traits<G_t>::edge_descriptor edge_descriptor;
    typedef typename boost::graph_traits<G_t>::vertex_descriptor vertex_descriptor;
    virtual ~graph_callback(){};
    virtual void operator()(vertex_descriptor)=0;
    virtual void operator()(vertex_descriptor, vertex_descriptor)=0;
};

} // treedec

#endif
