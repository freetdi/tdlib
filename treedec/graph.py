# (c) 2021 Felix Salfelder
# GPLv3+

from . import _graph
from ._graph import _balvvd
