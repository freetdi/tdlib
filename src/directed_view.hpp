// Felix Salfelder 2017
//
// (c) 2017 Felix Salfelder
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//
#ifndef TDI_DIRECTED_VIEW_H
#define TDI_DIRECTED_VIEW_H

#include <boost/graph/graph_traits.hpp>
#include <boost/graph/adjacency_list.hpp>
// #include <boost/graph/copy.hpp> BUG
#include "treedec_traits.hpp"

#ifdef USE_GALA
#include <gala/boost.h>
#endif

namespace treedec{

namespace draft{

namespace detail{

template<class G, class X=void>
struct dwt{
	typedef typename graph_traits<G>::directed_type type;

	static std::string dbg(){ return "wrap directed\n"; }

	static size_t init(G const& g){
		return boost::num_vertices(g);
	}

	template<class GG, class H>
	static void copy(GG const& g, H& h){
		assert(boost::is_undirected(g));
		auto p=boost::edges(g);
		for(; p.first!=p.second; ++p.first){
			auto V=boost::source(*p.first, g);
			auto W=boost::target(*p.first, g);
			boost::add_edge(V, W, h);
			boost::add_edge(W, V, h);
		}
		trace2("", boost::num_edges(g), boost::num_edges(h));
		assert(boost::num_edges(g)*2 == boost::num_edges(h));
	}
};

template<class G>
struct dwt<G,
	typename std::enable_if<
	   std::is_convertible< typename boost::graph_traits<G>::directed_category,
	                        boost::directed_tag>::value
	    && !std::is_convertible< typename boost::graph_traits<G>::directed_category,
	                             boost::bidirectional_tag>::value, void >::type >
{ //
	typedef typename graph_traits<G>::directed_type& type;
	static std::string dbg(){ return "dummy wrapper\n"; }

	static G& init(G& g){
		return g;
	}

	template<class GG, class H>
	static void copy(GG const&, H&){
	}
};

#ifdef USE_GALA // for now.
VCTtemplate
struct dwt< gala::graph<SGARGS>,
	typename std::enable_if<
	 std::is_convertible<
	   typename boost::graph_traits<gala::graph<SGARGS> >::directed_category,
		boost::directed_tag>::value
	 && !std::is_convertible<
	   typename boost::graph_traits<gala::graph<SGARGS> >::directed_category,
		boost::bidirectional_tag>::value ,
	 void >::type >
{ //
	typedef gala::graph<SGARGS> G;
	typedef typename gala::graph<SGARGS>::directed_type type;
	static std::string dbg(){ return "gala wrapper\n"; }

	static type init(G&){ untested();
		return type();
	}

	template<class GG, class H>
	static void copy(GG && g, H& h){ untested();
		h = std::move(g);
	}
};
#else
#endif
template<class G>
struct dwt<G,
	typename std::enable_if< std::is_convertible<
	typename boost::graph_traits<G>::directed_category,
  	boost::bidirectional_tag>::value, void >::type >
{ //
	typedef typename graph_traits<G>::directed_type type;
	static std::string dbg(){ return "bidir wrapper\n"; }

	static size_t init(G&){
		return 0;
	}

	// check: do we need this copy?
	template<class GG, class H>
	static void copy(GG const& g, H& h){
		assert(!boost::num_vertices(h));
		assert(boost::is_directed(h));
		auto p=boost::edges(g);
		for(; p.first!=p.second; ++p.first){
			auto e=*p.first;
			auto V=boost::source(e, g);
			auto W=boost::target(e, g);
			boost::add_edge(V, W, h);
			boost::add_edge(W, V, h);
			trace2("bidir cp", W, V);
		}
		trace2("bidir cp", boost::num_edges(g), boost::num_edges(h));
		assert(2*boost::num_edges(g) == boost::num_edges(h));
	}
};

}

template<class G>
class directed_view{
public:
	typedef detail::dwt<G> wrapper_help;
	typedef typename detail::dwt<G>::type backend_type;
	typedef typename std::remove_reference<backend_type>::type wrapped_type;

	typedef typename boost::graph_traits<wrapped_type> wrapped_traits;
	typedef typename wrapped_traits::vertex_descriptor vertex_descriptor;
	typedef typename wrapped_traits::edge_descriptor edge_descriptor;
	typedef typename wrapped_traits::adjacency_iterator adjacency_iterator;
	typedef typename wrapped_traits::vertex_iterator vertex_iterator;
	typedef typename wrapped_traits::edge_iterator edge_iterator;
	typedef typename wrapped_traits::out_edge_iterator out_edge_iterator;
	typedef typename wrapped_traits::vertices_size_type vertices_size_type;
	typedef typename wrapped_traits::degree_size_type degree_size_type;
	typedef typename wrapped_traits::edges_size_type edges_size_type;

	typedef typename wrapped_traits::edge_parallel_category edge_parallel_category;

	typedef typename boost::directed_tag directed_category;
	typedef typename boost::adjacency_graph_tag traversal_category;

private:
	directed_view(){ unreachable(); }
	directed_view(const directed_view& ) { unreachable();}
public:
	directed_view(G& g, bool commit=false)
	 : _g(wrapper_help::init(g)),
	   _commit(commit)
	{
		assert(boost::is_directed(_g));

		// no, only copies one edge per edge
		//boost::copy_graph(g, _g);
		//
		trace1("wrapping", wrapper_help::dbg());

		wrapper_help::copy(g, _g);
	}

	~directed_view(){
		if(_commit){ untested();
			incomplete();
		}else{
		}
	}

public: // dangerous?
	wrapped_type* operator->(){ untested();
		return &_g;
	}
	wrapped_type& operator*(){ itested();
		return _g;
	}
	wrapped_type const& operator*() const{ itested();
		return _g;
	}
public: // unnecessary
	std::pair< typename boost::graph_traits<G>::vertex_iterator,
		typename boost::graph_traits<G>::vertex_iterator>
	vertices() const{ untested();
		return boost::vertices(_g);
	}
	std::pair<adjacency_iterator, adjacency_iterator>
	adjacent_vertices(vertex_descriptor v) const{
		return boost::adjacent_vertices(v, _g);
	}
	std::pair<out_edge_iterator, out_edge_iterator>
	out_edges(vertex_descriptor v) const{
		return boost::out_edges(v, _g);
	}
	typename boost::graph_traits<G>::vertices_size_type
	degree(typename boost::graph_traits<G>::vertex_descriptor v) const{ untested();
      wrapped_type* g = const_cast<wrapped_type*>(&_g);
		return boost::out_degree(v, *g);
	}
private:
	backend_type _g;
	bool _commit;
};

} // draft

} // treedec

namespace boost{

template<class G>
struct graph_traits<treedec::draft::directed_view<G> > {
	typedef treedec::draft::directed_view<G> B;
	typedef typename treedec::draft::directed_view<G>::wrapped_type WW;
	typedef typename std::remove_reference<WW>::type W;

	typedef typename B::vertices_size_type vertices_size_type;
	typedef typename B::degree_size_type degree_size_type;
	typedef typename B::edges_size_type edges_size_type;

	typedef typename B::vertex_descriptor vertex_descriptor;
	typedef typename B::edge_descriptor edge_descriptor;

	typedef typename B::vertex_iterator vertex_iterator;
	typedef typename B::edge_iterator edge_iterator;
	typedef typename B::out_edge_iterator out_edge_iterator;
	typedef typename B::directed_category directed_category;
	typedef typename B::adjacency_iterator adjacency_iterator;

	// hmm.
	// typedef typename boost::disallow_parallel_edge_tag edge_parallel_category;
	typedef typename B::edge_parallel_category edge_parallel_category;

	typedef typename B::traversal_category traversal_category;

	static vertex_descriptor null_vertex(){ untested();
		return 0; // graph_traits<W>::null_vertex();
	}
};

template<class G, class T>
struct property_map<treedec::draft::directed_view<G>, T> {
	typedef typename property_map<
		typename treedec::draft::directed_view<G>::wrapped_type, T>::const_type const_type;
	typedef typename property_map<
		typename treedec::draft::directed_view<G>::wrapped_type, T>::type type;
};

template<class G>
degree_property_map<treedec::draft::directed_view<G> >
get(vertex_degree_t t, treedec::draft::directed_view<G> const& g)
{ untested();
	return get(t, *g);
}

template<class G>
degree_property_map<treedec::draft::directed_view<G> >
get(vertex_degree_t t, treedec::draft::directed_view<G>& g)
{ untested();
	return get(t, *g);
}

template<class T, class G>
typename property_map<typename treedec::draft::directed_view<G>::wrapped_type,
                      T>::type
get(T t, treedec::draft::directed_view<G>& g)
{
	return get(t, *g);
}

#if 1
template<class G>
typename treedec::draft::directed_view<G>::vertices_size_type
get(vertex_index_t,
		treedec::draft::directed_view<G> const& g,
		typename treedec::draft::directed_view<G>::vertex_descriptor v
		)
{ itested();
	return get(get(vertex_index, *g), v);
}
#endif

template<class T, class G>
typename property_map<typename treedec::draft::directed_view<G>::wrapped_type,
                      T>::const_type
get(T t, treedec::draft::directed_view<G> const& g)
{
	return get(t, *g);
}

template<class T, class G, class V>
typename property_map<typename treedec::draft::directed_view<G>::wrapped_type,
                      T>::value_type
get(T t, treedec::draft::directed_view<G> const& g, V v)
{ untested();
	return get(t, *g, v);
}

template<class T, class G, class V, class W>
void put(T t, treedec::draft::directed_view<G>& g, V v, W w)
{ untested();
	return put(t, *g, v, w);
}

template<class G>
typename property_map<typename treedec::draft::directed_view<G>::wrapped_type,
                      boost::edge_all_t>::const_type
get(edge_all_t, treedec::draft::directed_view<G> const& g)
{
	return get(edge_all, *g);
}

template<class G>
typename treedec::draft::directed_view<G>::vertices_size_type
num_vertices(treedec::draft::directed_view<G> const& g)
{
	return num_vertices(*g);
}

template<class G>
typename graph_traits<G>::edges_size_type
num_edges(treedec::draft::directed_view<G> const& g)
{
	untested();
	return num_edges(*g);
}

template<class VD, class G>
typename graph_traits<G>::vertices_size_type
out_degree(VD v, treedec::draft::directed_view<G> const& g)
{
	return out_degree(v, *g);
}

template<class VD, class G>
typename graph_traits<G>::vertices_size_type
	// BUG: out_degree!
degree(VD v, treedec::draft::directed_view<G> const& g)
{
	return out_degree(v, *g);
}

template<class VD, class G>
void clear_vertex(VD v, treedec::draft::directed_view<G>& g)
{
	return clear_vertex(v, *g);
}

template<class VD, class G>
std::pair<typename treedec::draft::directed_view<G>::edge_descriptor, bool >
add_edge(VD v, VD w, treedec::draft::directed_view<G>& g)
{
#ifndef NDEBUG
	// hmm directed_view disallows multiedges.
	if(edge(v, w, *g).second){ untested();
		incomplete();
		// that will not work with vectors
		return std::make_pair(edge(v, w, *g).first, true);
	}
#endif
	return add_edge(v, w, *g);
}

template<class G>
void remove_edge(
		typename treedec::draft::directed_view<G>::vertex_descriptor v,
		typename treedec::draft::directed_view<G>::vertex_descriptor w,
		treedec::draft::directed_view<G>& g)
{ untested();
	return remove_edge(v, w, *g);
}

template<class G, class P>
void remove_out_edge_if(
		typename treedec::draft::directed_view<G>::vertex_descriptor v,
		P p, treedec::draft::directed_view<G>& g)
{ untested();
	return remove_out_edge_if(v, p, *g);
}

template<class G>
std::pair<typename treedec::draft::directed_view<G>::edge_descriptor, bool >
edge(
	typename treedec::draft::directed_view<G>::vertex_descriptor v,
	typename treedec::draft::directed_view<G>::vertex_descriptor w,
	treedec::draft::directed_view<G>& g)
{
	return edge(v, w, *g);
}

template<class G>
std::pair<typename treedec::draft::directed_view<G>::edge_descriptor, bool >
edge(
	typename treedec::draft::directed_view<G>::vertex_descriptor v,
	typename treedec::draft::directed_view<G>::vertex_descriptor w,
	treedec::draft::directed_view<G> const& g)
{ untested();
	return edge(v, w, *g);
}

template<class G>
typename treedec::draft::directed_view<G>::vertex_descriptor
source(typename treedec::draft::directed_view<G>::edge_descriptor w,
		treedec::draft::directed_view<G> const& g)
{
	return source(w, *g);
}

// bug: disable if edge_descriptor is a pair?!
template<class G>
typename treedec::draft::directed_view<G>::vertex_descriptor
target(typename treedec::draft::directed_view<G>::edge_descriptor w,
		treedec::draft::directed_view<G> const& g)
{
	return target(w, *g);
}

template<class G>
std::pair<
	typename treedec::draft::directed_view<G>::vertex_iterator,
	typename treedec::draft::directed_view<G>::vertex_iterator>
vertices(treedec::draft::directed_view<G> const& g)
{
	return vertices(*g);
}

template<class G>
std::pair<
	typename treedec::draft::directed_view<G>::edge_iterator,
	typename treedec::draft::directed_view<G>::edge_iterator>
edges(treedec::draft::directed_view<G> const& g)
{
	return edges(*g);
}

template<class G>
std::pair<
	typename treedec::draft::directed_view<G>::out_edge_iterator,
	typename treedec::draft::directed_view<G>::out_edge_iterator>
out_edges(
		typename treedec::draft::directed_view<G>::vertex_descriptor v,
		treedec::draft::directed_view<G> const& g)
{
	return g.out_edges(v);
}

template<class G>
std::pair<
	typename treedec::draft::directed_view<G>::adjacency_iterator,
	typename treedec::draft::directed_view<G>::adjacency_iterator>
adjacent_vertices(
		typename treedec::draft::directed_view<G>::vertex_descriptor v,
		treedec::draft::directed_view<G> const& g)
{
	return g.adjacent_vertices(v);
}

    template<class G>
    struct vertex_bundle_type<treedec::draft::directed_view<G> > {
      typedef typename vertex_bundle_type<G>::type type;
    };

} // boost

namespace treedec {

template<class G>
struct directed_view_select{
    typedef treedec::draft::directed_view<G> type;
};

}

#endif // guard
