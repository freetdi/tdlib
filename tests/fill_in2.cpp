
#define DEBUG_FILL

#include <boost/graph/graphviz.hpp>
#include <boost/property_map/property_map.hpp>
#include <boost/graph/graph_traits.hpp>
#include "graph_traits.hpp"
// #include "treedec_traits.hpp"

#include <gala/cbset.h>
#include <gala/td.h>
#include <gala/boost.h>
#include <treedec/directed_view.hpp>


#include "induced_supergraph.hpp"
#include "elimination_orderings.hpp"
#include <gala/graph.h>

#include <stdarg.h>
#include <stdio.h>

#define assert_symmetric(g) { \
    unsigned i=0; \
    auto E=boost::edges(g); \
    for(;E.first!=E.second; ++E.first){ \
        ++i;\
        assert( boost::edge(boost::source(*E.first, g), \
                    boost::target(*E.first, g),g).second); \
        assert( boost::edge(boost::target(*E.first, g), \
                    boost::source(*E.first, g),g).second); \
    }\
    trace1("symmetric", i); \
}

template<class X, class ... rest>
struct algo_config : treedec::algo::default_config<X, rest...>{
    static void message(int badness, const char* fmt, ...) { itested();
        if (badness >= bLOG){
            char buffer[2048] = "c ";
            va_list arg_ptr;
            va_start(arg_ptr,fmt);
            vsprintf(buffer+2,fmt,arg_ptr);
            va_end(arg_ptr);
            std::cout << buffer;
        }else{
        }
    }
};

template<class G>
struct uvv_config : gala::graph_cfg_default<G> {
    static constexpr bool is_directed=false; // TODO: add "false" test.
    static constexpr bool force_simple=true;
    static constexpr bool force_loopless=true; // not used yet?
    // static constexpr bool force_symmetric=true; // meaninngless (undirected)
    // typedef tdDEGS<G> degs_type; // obsolete.
};
//typedef gala::graph<std::vector, std::vector, unsigned, uvv_config> graph_t;
typedef gala::graph<std::vector, std::vector, unsigned> graph_t;
typedef typename treedec::graph_traits<graph_t>::treedec_type T;

typedef treedec::pending::impl::fillIn<graph_t, algo_config> algo_type;

template<class A, class G>
void biedge(A a, A b, G& g)
{
	boost::add_edge(a,b,g);
	boost::add_edge(b,a,g);
}

void clique(int n)
{
	graph_t g(n);

	std::vector<std::pair<int, int>> edges;

	for (int x = 0; x < n; x++) {
		for (int y = 0; y < x; y++) {
			biedge(x, y, g);
		}
	}
	assert_symmetric(g);

	auto i = boost::edges(g);
	auto k=0;
	for(; i.first!=i.second; ++i.first){
		++k;
		trace2("clique edg", boost::source(*i.first, g), boost::target(*i.first, g));
	}
	trace2("dbg", k, n);
	assert(k == (n-1) * n / 2);

//	auto j = boost::adjacent_vertices(4, g);
//	for(; j.first!=j.second; ++j.first){
//		trace1("grid", *j.first);
//	}

	auto vi = boost::vertices(g);
	for(; vi.first!=vi.second; ++vi.first){
		unsigned d = boost::degree(*vi.first, g);
		trace2("clique:", *vi.first, d);
	}

	algo_type a(g);

	a.do_it();

	T t;
	a.get_tree_decomposition(t);

	if(treedec::get_bagsize(t) == a.get_bagsize()){
	}else{
		incomplete();
	}
	assert(treedec::is_valid_treedecomposition(g, t));
	std::cout << "done clique " << n << " fi " << a.get_bagsize() << "\n";
}

int main(int, char**)
{
	clique(1);
	clique(2);
	clique(3);
	clique(6);
	clique(10);
	clique(15);
	clique(20);
}
