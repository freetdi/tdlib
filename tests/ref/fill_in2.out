c done. max fill=0
done clique 1 fi 1
c merge 0 (1) into 1 (1)
c next 1. degree 1, fill 0, sns 2
c done. max fill=0
done clique 2 fi 2
c merge 0 (1) into 2 (1)
c merge 1 (1) into 2 (2)
c next 2. degree 2, fill 0, sns 3
c done. max fill=0
done clique 3 fi 3
c merge 0 (1) into 5 (1)
c merge 1 (1) into 5 (2)
c merge 2 (1) into 5 (3)
c merge 3 (1) into 5 (4)
c merge 4 (1) into 5 (5)
c next 5. degree 5, fill 0, sns 6
c done. max fill=0
done clique 6 fi 6
c merge 0 (1) into 9 (1)
c merge 1 (1) into 9 (2)
c merge 2 (1) into 9 (3)
c merge 3 (1) into 9 (4)
c merge 4 (1) into 9 (5)
c merge 5 (1) into 9 (6)
c merge 6 (1) into 9 (7)
c merge 7 (1) into 9 (8)
c merge 8 (1) into 9 (9)
c next 9. degree 9, fill 0, sns 10
c done. max fill=0
done clique 10 fi 10
c merge 0 (1) into 14 (1)
c merge 1 (1) into 14 (2)
c merge 2 (1) into 14 (3)
c merge 3 (1) into 14 (4)
c merge 4 (1) into 14 (5)
c merge 5 (1) into 14 (6)
c merge 6 (1) into 14 (7)
c merge 7 (1) into 14 (8)
c merge 8 (1) into 14 (9)
c merge 9 (1) into 14 (10)
c merge 10 (1) into 14 (11)
c merge 11 (1) into 14 (12)
c merge 12 (1) into 14 (13)
c merge 13 (1) into 14 (14)
c next 14. degree 14, fill 0, sns 15
c done. max fill=0
done clique 15 fi 15
c merge 0 (1) into 19 (1)
c merge 1 (1) into 19 (2)
c merge 2 (1) into 19 (3)
c merge 3 (1) into 19 (4)
c merge 4 (1) into 19 (5)
c merge 5 (1) into 19 (6)
c merge 6 (1) into 19 (7)
c merge 7 (1) into 19 (8)
c merge 8 (1) into 19 (9)
c merge 9 (1) into 19 (10)
c merge 10 (1) into 19 (11)
c merge 11 (1) into 19 (12)
c merge 12 (1) into 19 (13)
c merge 13 (1) into 19 (14)
c merge 14 (1) into 19 (15)
c merge 15 (1) into 19 (16)
c merge 16 (1) into 19 (17)
c merge 17 (1) into 19 (18)
c merge 18 (1) into 19 (19)
c next 19. degree 19, fill 0, sns 20
c done. max fill=0
done clique 20 fi 20
